### ISSUE TITLE
bug: Stencil first prod build is broken after dev build

### ISSUE THREAD

**Issue by cmaas** [ionic-team/stencil#4948]
**Update 2023-10-25:** The first Stencil prod build after a dev build always seem to differ from the second (or third...) build. Actually, has nothing to do with Ionic. In the first build, the whole `patchBrowser = () => ...` stuff is present (non-minified), whereas a second or third build creates a different kind of release file. But if you include Ionic, the whole app doesn't even load, because of stuff like this: `import './ionic-global-8c59e7e6.js';`.

Still, I don't think the first prod build should differ from subsequent builds at all. Quite confusing and error-prone.

### Prerequisites

- [X] I have read the [Contributing Guidelines](https://github.com/ionic-team/stencil/blob/main/CONTRIBUTING.md).
- [X] I agree to follow the [Code of Conduct](https://github.com/ionic-team/stencil/blob/main/CODE_OF_CONDUCT.md).
- [X] I have searched for [existing issues](https://github.com/ionic-team/stencil/issues) that already report this problem, without success.

### Stencil Version

4.5.0

### Current Behavior

```
"@ionic/core": "7.5.1",
"@stencil/core": "4.5.0"
```

A project that uses Ionic and Stencil creates a **broken build** when running `stencil build` for the first time after `stencil build --dev --watch --serve`. I always have to build twice to get the proper output.

The output of the first run contains a JS file that includes stuff that simply doesn't exist:

`index.html`:
```html
<script type="module" src="/build/p-630a4e0e.js" data-stencil data-resources-url="/build/"
        data-stencil-namespace="app"></script>
```

`/build/p-630a4e0e.js` (excerpt):
```js
import { B as BUILD, c as consoleDevInfo, H, d as doc, N as NAMESPACE, p as promiseResolve, b as bootstrapLazy } from './index-5be4bbac.js';
export { s as setNonce } from './index-5be4bbac.js';
import { g as globalScripts } from './app-globals-f8ab0fed.js';
import './ionic-global-8c59e7e6.js';
import './utils-23c30c62.js';
import './animation-4811d121.js';
import './index-150d1a25.js';
import './helpers-a0b2a380.js';
import './index-914e555f.js';
import './ios.transition-1739a486.js';
import './md.transition-538de680.js';
import './cubic-bezier-67bf1f69.js';
import './index-89e0f6ee.js';
import './config-a7320371.js';
import './theme-7dce6eff.js';
import './index-d0e75c8c.js';
import './index-bf616dc7.js';
import './hardware-back-button-90902d55.js';
import './overlays-1cb3af53.js';
import './framework-delegate-0227fbeb.js';

/*
 Stencil Client Patch Browser v4.5.0 | MIT Licensed | https://stenciljs.com
 */
const patchBrowser = () => {
    // NOTE!! This fn cannot use async/await!
    if (BUILD.isDev && !BUILD.isTesting) {
        consoleDevInfo('Running in development mode.');
    }
//...
```

Most of these imported files are not present in the actual build output. If `stencil build` is run a second time, a different JS file is created:

```html
<script type="module" src="/build/p-3040ee01.js" data-stencil data-resources-url="/build/"
        data-stencil-namespace="app"></script>
```

Which looks roughly like this `/build/p-3040ee01.js` (first few lines): 
```js
import{p as e,b as o}from"./p-ef0d46fc.js";export{s as setNonce}from"./p-ef0d46fc.js";import{g as n}from"./p-3bebd51e.js";import"./p-f38c8989.js";import"./p-fd141259.js";const t=()=>{const o=import.meta.url;const n={};if(o!==""){n.resourcesUrl=new URL(".",o).href}return e(n)};t().then((e=>{n();return o(JSON.parse('[["p-f5d63535",[[33,
//...
```

If you deploy the broken build, the app cannot start, because it fails to load non-existing .js files.

I disabled the cache in `stencil.config.ts`, but it doesn't make a difference. I provided a sample Github repo below, but it is essentially the Stencil starter with `ionic-pwa` and updated all dependencies to the most recent versions.

### Expected Behavior

Build the proper prod build the first time when I run `stencil build`.

### System Info

```shell
System: node 20.7.0
    Platform: darwin (22.5.0)
   CPU Model: Apple M1 Max (10 cpus)
    Compiler: /Users/cm/Code/ionic-bugtest/node_modules/@stencil/core/compiler/stencil.js
       Build: 1697469010
     Stencil: 4.5.0 📢
  TypeScript: 5.2.2
      Rollup: 2.42.3
      Parse5: 7.1.2
      Sizzle: 2.42.3
      Terser: 5.21.0
```

### Steps to Reproduce

1. Take any Stencil + Ionic project, run the default npm start command, which is `stencil build --dev --watch --serve`
2. Cancel the watcher, run the default npm run build, which is `stencil build` to create a prod build
3. Try to load the resulting app in your browser or simply inspect the files in `www/`

### Code Reproduction URL

https://github.com/cmaas/ionic-bugtest

### Additional Information

- This has been an issue for years and I remember that it was pointed out before, but can't find it anymore.
- At least one other person reported the same problem in the Discord.

**Comment by cmaas** [comment]
Tried something: The first Stencil prod build after a dev build always seem to differ from the second (or third...) build. Actually, has nothing to do with Ionic. In the first build, the whole `patchBrowser = () => ...` stuff is present (non-minified), whereas a second or third build creates a different kind of release file. But if you include Ionic, the whole app doesn't even load, because of stuff like this: `import './ionic-global-8c59e7e6.js';`.

Including Ionic only makes this problem visible. Still, I don't think the first prod build should differ from subsequent builds at all. Quite confusing and error-prone.

**Comment by alicewriteswrongs** [comment]
Hey @cmaas, I've just been taking a look at this as well and was able to reproduce the issue both with your reproduction case (thanks for providing that!) as well as in a plain Stencil project I created with `npm init stencil`. You're definitely right that there's no good reason the subsequent builds should be different...I'll dig into this and figure out what's going on!

**Comment by alicewriteswrongs** [comment]
@cmaas alright so I think I've gotten to the bottom of this, it's, as always, a race condition where the WWW output target depends on a given file already being present on disk and generally speaking it isn't _but_ since output from previous builds sticks around in the output directory it will see a file from the previous build, think it should be included in the current build, and some weird stuff can happen as a result.

I have a fix which I think will eliminate the issue, and I just put together a dev build for it. Would you mind trying it and seeing if it fixes your issue? You can install it like so:

```sh
npm i @stencil/core@4.6.0-dev.1698266136.5552fc1
```

let me know if that works and if so I'll get a PR up!

**Comment by ptmdmusique** [comment]
@alicewriteswrongs 👋 thanks for the update! Unfortunately, I think this also happens with the `component.d.ts` file where the first `npm run build` always fails if we use `@Element() el: HTMLMyComponentElement;` inside the same component

I _think_ this might be because it tries to check for type error but `component.d.ts` isn't available yet. That's why running `npm run build` the 2nd time "fixes" it. Maybe some race condition is happening behind the scene ❓ 

Here is the repo ... repo
https://github.com/ptmdmusique/stencil-build-bug

**Comment by cmaas** [comment]
@alicewriteswrongs Thanks for looking into this. Yes, `4.6.0-dev.1698266136.5552fc1` seems to resolve this. I could achieve the same result if I change my package.json to remove the entire build directory first like so:
```json
"scripts": {
    "build": "rm -r www; stencil build",
    "start": "rm -r www; stencil build --dev --watch --serve"
}
```

This way, the first prod build always starts with a clean slate.

@ptmdmusique Isn't that by design? Your code relies on a type (`HTMLMyComponentElement`) that you didn't create yourself. Wouldn't that type definition be more useful for the component consuming `HTMLMyComponentElement`? Although it's indeed weird that the first build fails, but the second doesn't. From my understanding, the build should always fail, because it relies on code generated by its own build output.

**Comment by alicewriteswrongs** [comment]
@cmaas glad to hear it fixes the issue! I'm putting up a PR now and hopefully I'll merge the fix today so it will be in the next release of Stencil (and in nightly builds starting the night after whenever it lands).

It makes sense that clearing out `www` fixes the issue - essentially what's going on it that the `www` build depends on a file that is produced by the lazy output target but it seems that quite consistently this file hasn't yet been written to disk when the `www` build tries to access it. So this strange inconsistency between builds is caused by the `www` build picking up the file from the _last_ build and treating it as if it's from the current build. Not good! But anyway, obviously if that's the root cause then clearing out the old build between runs should fix it.

@ptmdmusique I think I agree with @cmaas here, it seems like you're introducing a circular dependency where the source code depends on a result of it's own compilation. One solution to this would I think be checking the `components.d.ts` file into version control -- this is the workflow [we recommend for this file](https://stenciljs.com/docs/typed-components).

**Comment by ptmdmusique** [comment]
Thanks guys! That's clear to me 👏 I guess there is no real use case for `@Element ele: MyComponent` and `@Element ele: HTMLElement` is enough since we can just access `this` to get/set the tag's attributes

I updated the [test case](https://github.com/ptmdmusique/stencil-build-bug/blob/master/src/components/my-also-component/my-also-component.tsx#L6) a bit based on the workflow you suggested and seems like it will fail since the type isn't available on first compilation. checking `components.d.ts` into version control will definitely solve this but seems like that fail is expected for the first time build ❓

**Comment by alicewriteswrongs** [comment]
Hey @cmaas, I just merged a change (#4992) which should address the issue. It will be included in the next release of Stencil, so stay tuned for that!

**Comment by cmaas** [comment]
@alicewriteswrongs Thanks for the quick fix, good writeup and overall work on Stencil and Ionic. I'm a happy user since 2017 :)

**Comment by rwaskiewicz** [comment]
The fix for this issue has been released as a [part of today's Stencil v4.7.0 release](https://github.com/ionic-team/stencil/releases/tag/v4.7.0). Should the issue reappear, please feel free to open a new ticket. Thanks!
